windows-agent/internal/proservices/landscape/executor.go

```diff
@@ -189,3 +197,101 @@ func (e executor) setDefault(ctx context.Context, cmd *landscapeapi.Command_SetD
 func (e executor) shutdownHost(ctx context.Context, cmd *landscapeapi.Command_ShutdownHost) error {
 	return gowsl.Shutdown(ctx)
 }
+
+func installFromMicrosoftStore(ctx context.Context, distro gowsl.Distro) (err error) {
+	defer decorate.OnError(&err, "can't install from Microsoft Store")
+
+	if err := gowsl.Install(ctx, distro.Name()); err != nil {
+		return err
+	}
+
+	if err := distroinstall.InstallFromExecutable(ctx, distro); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func installFromURL(ctx context.Context, homeDir string, distro gowsl.Distro, rootfs *landscapeapi.Command_Install_Rootfs) (err error) {
+	defer decorate.OnError(&err, "can't install from URL: %q", rootfs.GetUrl())
+
+	tmpDir := filepath.Join(os.TempDir(), distro.Name())
+	if err := os.MkdirAll(tmpDir, 0700); err != nil {
+		return err
+	}
+	// Remove tarball once installed
+	defer os.RemoveAll(tmpDir)
+
+	tarball := filepath.Join(tmpDir, distro.Name()+".tar.gz")
+
+	f, err := os.Create(tarball)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	err = download(ctx, f, rootfs.GetUrl(), rootfs.GetSha256Sum())
+	if err != nil {
+		return err
+	}
+
+	// Create the directory that will contain the vhdx
+	vhdxDir := filepath.Join(homeDir, "WSL", distro.Name())
+	if err := os.MkdirAll(vhdxDir, 0700); err != nil {
+		return err
+	}
+
+	if _, err := gowsl.Import(ctx, distro.Name(), tarball, vhdxDir); err != nil {
+		err = os.RemoveAll(vhdxDir)
+		if err != nil {
+			log.Warningf(ctx, "could not cleanup install directory: %v", err)
+		}
+		return err
```

Comment:
Now if both `gowsl.Import` and `os.RemoveAll` fail, we'll loose the error information on the first. I'd rather name the second err variable to something else and return `errors.Join`

Reply:
Ah yeah, an oversight on my part. Good spot :)